Add PartitionTable

tillrohrmann opened the pull request:
The partition table is the more flexible replacement of the FixedPartitionTable
that allows to assign different nodes to a partition and change leadership.

This fixes https://github.com/restatedev/restate/issues/1701.

AhmedSoliman commented on crates/types/src/partition_table.rs:
```diff
@@ -93,21 +86,120 @@ impl FixedPartitionTable {
+#[derive(Debug, Clone, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct Partition {
+    pub key_range: RangeInclusive<PartitionKey>,
+    pub leader: Option<GenerationalNodeId>,
+    pub node_set: BTreeSet<GenerationalNodeId>,
```
I'm not sure whether those should exist here. I think operational status and deployment configuration for partitions should be store separately (those are cluster controller's metadata)

In the future we will need to have a `state` for the partition so we can archive old partitions which won't appear in the index but can still be addressed by partition_id. But that's something we can add when we need it. If you'd like to take the first step in this, maybe we can have an enum with `Active` / `Archived` states.

tillrohrmann replied:
The idea behind the `leader` field was that maybe some components like the `ingress` would like to know who the current leader is for a given partition to query that leader (e.g. when submitting an invocation or retrieving an invocation result). The `node_set` is not so helpful, agreed. I can remove both fields until we have a concrete need for it.

tillrohrmann commented on crates/types/src/partition_table.rs:
```diff
@@ -38,45 +38,38 @@ impl From<KeyRange> for RangeInclusive<PartitionKey> {
-pub struct FixedPartitionTable {
+#[serde(try_from = "PartitionTableShadow", into = "PartitionTableShadow")]
+pub struct PartitionTable {
     version: Version,
-    num_partitions: u64,
+    partitions: BTreeMap<PartitionId, Partition>,
+    partition_key_index: BTreeMap<PartitionKey, PartitionId>,
```
Good point. I will add it.

tillrohrmann commented on crates/types/src/partition_table.rs:
```diff
@@ -93,21 +86,120 @@ impl FixedPartitionTable {
+
+#[serde_with::serde_as]
+#[derive(serde::Serialize, serde::Deserialize)]
+struct PartitionTableShadow {
```
True. Especially myself.

tillrohrmann commented on crates/types/src/partition_table.rs:
```diff
@@ -93,21 +86,120 @@ impl FixedPartitionTable {
+    #[serde_as(as = "Option<serde_with::Seq<(_, _)>>")]
+    partitions: Option<BTreeMap<PartitionId, Partition>>,
+    // flexbuffers only supports string-keyed maps :-( --> so we store it as vector of kv pairs
+    #[serde_as(as = "Option<serde_with::Seq<(_, _)>>")]
+    partition_key_index: Option<BTreeMap<PartitionKey, PartitionId>>,
```
Creating this field on deserialization sounds like a better solution. Will change it.

tillrohrmann commented on crates/types/src/partition_table.rs:
```diff
@@ -93,21 +86,120 @@ impl FixedPartitionTable {
+                partitions,
+                partition_key_index,
+            }),
+            (None, None) => Ok(PartitionTable::with_equally_sized_partitions(
```
Yes, exactly. Will leave a comment to clarify it.

tillrohrmann commented on crates/types/src/partition_table.rs:
```diff
@@ -93,21 +86,120 @@ impl FixedPartitionTable {
+#[derive(serde::Serialize, serde::Deserialize)]
+struct PartitionTableShadow {
+    version: Version,
+    // only needed for the FixedPartitionTable
```
True. Will change it.

tillrohrmann replied:
Technically, we would need a way to make sure that there are no more `FixedPartitionTables` before we can remove this bit. Right now, when upgrading from v1 to v1.1 and if there is no update to the `PartitionTable`, we would keep the serialized data of the `FixedPartitionTable`.

tillrohrmann commented:
@AhmedSoliman I've pushed an updated version where I addressed all your comments except for the `leader` field in `Partition`. I am fine with removing it as premature/unneeded as well.

AhmedSoliman commented:
@tillrohrmann my thinking is that routing information should be live (by propagating ClusterState) and not relying on persisted metadata. So, I'd suggest we remove it.

tillrohrmann commented:
Alrighty. Makes sense. I'll remove it.

AhmedSoliman reviewed (approved):
Skimmed over it and it looks good.

tillrohrmann commented on crates/core/src/metadata/mod.rs:
```diff
@@ -127,7 +127,7 @@ impl Metadata {
+    ) -> Result<Arc<PartitionTable>, ShutdownError> {
         let partition_table = self.partition_table_ref();
         if partition_table.version() >= min_version {
             return Ok(partition_table.into_arc());
```
This would probably give a more flexible API because one can always convert the `Pinned` into an `Arc`. Will update the code.

tillrohrmann commented:
Thanks for the review @AhmedSoliman. I've addressed your comment and rebased on the latest `main`. Once Github gives green light, I will merge this PR.
